content/drop-stdlib-forwards-bin-compat.md

```diff
@@ -0,0 +1,295 @@
+---
+layout: sip
+permalink: /sips/:title.html
+stage: pre-sip
+status: submitted
+title: SIP-NN - Drop Forwards Binary Compatibility of the Scala 2.13 Standard Library
+---
+
+**By: Lukas Rytz**
+
+
+## History
+
+| Date           | Version            |
+|----------------|--------------------|
+| Dec 8, 2022    | Initial Version    |
+
+
+## Summary
+
+I propose to drop the forwards binary compatibility requirement that build tools enforce on the Scala 2.13 standard library.
+This will allow implementing performance optimizations of collection operations that are currently not possible.
+It also unblocks adding new classes and new members to existing classes in the standard library.
+
+
+## Backwards and Forwards Compatibility
+
+A library is backwards binary compatible if code compiled against an old version works with a newer version on the classpath.
+Forwards binary compatibility requires the opposite: code compiled against a new version of a library needs to work with an older version on the classpath.
+A more in-depth explanation of binary compatibility is available on the [Scala documentation site](https://docs.scala-lang.org/overviews/core/binary-compatibility-of-scala-releases.html).
+
+Scala build tools like sbt automatically update dependencies on the classpath to the latest patch version that any other dependency on the classpath requires.
+For example, with the following definition
+
+~~~ scala
+libraryDependencies ++= List(
+  "com.softwaremill.sttp.client3" %% "core" % "3.8.3", // depends on ws 1.3.10
+  "com.softwaremill.sttp.shared"  %% "ws"   % "1.2.7", // for demonstration
+)
+~~~
+
+sbt updates the `ws` library to version 1.3.10.
+Running the `evicted` command in sbt displays all dependencies whose version were changed.
+
+This build tool feature allows library authors to only maintain backwards binary compatibility in new versions, they don't need to maintain forwards binary compatibility.
+Backwards binary compatible changes include the addition of new methods in existing classes and the addition of new classes.
+Such additions don't impact existing code that was compiled against an older version, all definitions that were previously present are still there.
+
+### The Standard Library
+
+The Scala standard library is treated specially by sbt and other build tools, its version is always pinned to the `scalaVersion` of the build definition and never updated automatically.
+
+For example, the `"com.softwaremill.sttp.client3" %% "core" % "3.8.3"` library has a dependency on `"org.scala-lang" % "scala-library" % "2.13.10"` in its POM file.
+When a project uses this version of the sttp client in a project with `scalaVersion` 2.13.8, sbt will put the Scala library version 2.13.8 on the classpath.
+
+This means that the standard library is required to remain both backwards and forwards binary compatible.
```

Comment:
I feel like the relationship between the sbt's alignment behavior and Scala library's compatibility requirement is a bit overstated here. 

- The policy of compatibility predates the introduction of alignment in 2016.
- Scala 3 dropped forward compatibility.

Pre-2016, I think Scala 2.x kept forward compatibility, I think because _not_ keeping the forward compatibility means coloring every library with an effective minimum Scala version beyond 0th patch i.e. 2.11.0, 2.12.0, 2.13.0 etc. So if someone wanted to use Maven or just download a bunch of JARs manually and use Ant or something they could use any patch version.

Reply:
Note prior to sbt 1.3.0 (2019) sbt used Apache Ivy, which emulates Maven's nearest-wins dependency resolution semantics, so it may not have resolved to the latest scala-version found in the transitive graph.